rooms: move room reading and storage to TRX

lahm86 opened the pull request:
#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TRX/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change
- [x] I have added a readme entry about my new feature or OG bug fix, or it is a different change

#### Description

This replaces all instances of `g_RoomInfo[idx]` with `Room_Get(idx)`, replaces `g_RoomCount` with a getter in TRX, renames variables to a standard of `room` (mix of this and just `r` previously) and shifts room reading fully to TRX. I've re-organised the common level reader a little as the room mesh function is now private, so this part may be best viewed per-commit.

lahm86 committed:
tr1/misc: use Room_Get and standardize vars

This replaces all cases of direct g_RoomInfo access with Room_Get,
aside from level reading for now. Variables are also standardized to
room rather than simply r in some cases.

lahm86 committed:
tr2/misc: use Room_Get and standardize vars

This replaces all cases of direct g_Rooms access with Room_Get, aside
from level reading for now. Variables are also standardized to room
rather than simply r in some cases.

lahm86 committed:
rooms: remove global room count

This replaces the global room count with a getter/setter in TRX.

lahm86 committed:
rooms: move room storage to TRX

This removes the room globals and replaces with an accessor in TRX.

lahm86 committed:
viewport: make extern max X/Y functions

This makes max X/Y viewport functions available in TR2, to allow us to
move room reading fully to TRX later.

lahm86 committed:
effects: move effects const to TRX

Moves NO_EFFECT to TRX.

lahm86 committed:
level: read rooms in TRX

This moves room reading fully to TRX, including removing the external
facing floor data reader.

lahm86 committed:
level: make room mesh parsing private

This makes the room mesh parser private to the level reading module.

lahm86 committed:
level: introduce position and shade helpers

This introduces helper function for reading positions and shades.

rr- commented on src/libtrx/game/rooms/common.c:
```diff
@@ -96,6 +99,27 @@ static const int16_t *M_ReadTrigger(
+
+int32_t Room_GetTotalCount(void)
+{
+    return m_RoomCount;
```
Maybe we could do `Room_GetCount`?

lahm86 replied:
Done, thanks.

rr- commented on src/tr1/game/camera/common.c:
```diff
@@ -465,7 +467,8 @@ static void M_SmartShift(
     if (!bad_bottom) {
-        camera_box = r->sectors[z_sector + (x_sector + 1) * r->size.z].box;
+        camera_box =
+            room->sectors[z_sector + (x_sector + 1) * room->size.z].box;
```
It'd be nice to have a getter for this, eg `Room_Get???Sector???(ROOM *room, int32_t x_sector, int32_t z_sector)`. I can't think of a good naming as we have the more complex `Room_GetSector()` already that operates on a world pos.

lahm86 replied:
Yes, I think we can come up with something. I'll try it out

lahm86 replied:
We now have `Room_GetWorldSector` for getting a room's sector given X/Z world coords, and `Room_GetUnitSector` for getting one by X/Z index.

rr- commented on src/tr2/game/viewport.c:
```diff
@@ -202,3 +202,13 @@ void Viewport_AlterFOV(const int16_t view_angle)
+int32_t Viewport_GetMaxY(void)
+{
+    return g_PhdWinMaxY;
+}
```
While this change is fine, there's something I've been wanting to discuss with you. I'm not really satisfied with how the Viewport API is structured in both games in its present form. In TR1 we have a lot of getters, while in TR2, we're overly dependent on using global variables. My original concept for TR2 was to use a function like `Viewport_Get()` that would return a constant struct containing all the necessary information. But unfortunately the game relies on even more variables than TR1 so this struct turned out to be big and ugly. For modifying the viewport, there are a couple of scenarios to consider:
- resizing the window, of course
- changing the FOV during cutscenes (happens on every frame)
- changing the FOV during cinematic animations (happens on every frame)
- changing the FOV when the players tweaks their settings
- drawing a 3D pickup in the corner (in which case the viewport is temporarily changed)

Can you think of any design that we might use to improve this? It's difficult for me to estimate the effect of fusing some of the game variables together – and the whole purpose of the `game_vars` struct was to avoid handling this issue at the time :D

lahm86 replied:
Hmm, that's a tricky one as you say as sometimes making a struct or ton of getters/setters explodes the complexity. Maybe we can figure something out in the refactor though, happy to help look into it.

rr- commented on src/libtrx/game/level/common.c:
```diff
@@ -72,6 +94,74 @@ static void M_ReadFace3(FACE3 *const face, VFILE *const file)
+static void M_ReadRoomMesh(const int32_t room_num, VFILE *const file)
+{
+    ROOM *const room = Room_Get(room_num);
+    const INJECTION_MESH_META inj_data = Inject_GetRoomMeshMeta(room_num);
```
Having this dependency here is unfortunate – a better design would be to pass a dictionary of room number to injection mesh data via a parameter, like we do with `extra_pages` in `Level_ReadTexturePages`. But this would be way wordier.

lahm86 replied:
It's a bit of a hangover from TR1 injections, it'll likely change when we redo the format there. Potentially something passed to the module as a whole after inject_init with all the additional information we need here.

lahm86 committed:
rooms: simplify room count getter

lahm86 committed:
rooms: introduce utility sector getters

This provides a function to get a room's sector by world X/Z pos, and
another by X/Z index.

rr- reviewed (approved):
This is great stuff.
